core/src/main/java/com/linecorp/armeria/common/logging/PrometheusMetricRequestDecorator.java

```diff
@@ -0,0 +1,299 @@
+                    final String[] values = getLabelValues(log);
+                    metricAdapter.time(values, log.totalDurationNanos())
+                                 .responseBytes(values, log)
+                                 .decActiveRequests(values);
+
+                    if (isSuccess(log)) {
+                        metricAdapter.success(values);
+                    } else {
+                        metricAdapter.failure(values);
+                    }
+                }, RequestLogAvailability.COMPLETE);
+        return function.apply(ctx, req);
+    }
+
+    @Override
+    public O serve(Service<I, O> delegate, ServiceRequestContext ctx, I req) throws Exception {
+        return request(delegate::serve, ctx, req);
+    }
+
+    @Override
+    public O execute(Client<I, O> delegate, ClientRequestContext ctx, I req) throws Exception {
+        return request(delegate::execute, ctx, req);
+    }
+
+    /**
+     * Returns a new {@link Service} decorator that tracks request stats using the Prometheus metrics  library.
+     *
+     * @param collectorRegistry Wrapped Prometheus registry
+     * @param labels Labels names for metrics
+     * @param labelingFunction Must return a sorted map in natural order
+     * @param <I> Request type
+     * @param <O> Response type
+     * @return A service decorator function
+     */
+    public static <I extends Request, O extends Response> DecoratingServiceFunction<I, O> decorateService(
+            PrometheusRegistryWrapper collectorRegistry,
+            MetricLabel[] labels,
+            Function<RequestLog, Map<? extends MetricLabel, String>> labelingFunction) {
+        return new PrometheusMetricRequestDecorator<>(collectorRegistry,
+                                                      labels,
+                                                      labelingFunction,
+                                                      "armeria_server");
+    }
+
+    /**
+     * Returns a new {@link Client} decorator that tracks request stats using the Prometheus metrics  library.
+     *
+     * @param collectorRegistry Wrapped Prometheus registry
+     * @param labels Labels names for metrics
+     * @param labelingFunction map of labels to values
+     * @param <I> Request type
+     * @param <O> Response type
+     * @return A client decorator function
+     */
+    public static <I extends Request, O extends Response> DecoratingClientFunction<I, O> decorateClient(
+            PrometheusRegistryWrapper collectorRegistry,
+            MetricLabel[] labels,
+            Function<RequestLog, Map<? extends MetricLabel, String>> labelingFunction) {
+        return new PrometheusMetricRequestDecorator<>(collectorRegistry,
+                                                      labels,
+                                                      labelingFunction,
+                                                      "armeria_client");
+    }
+
+    private String[] getLabelValues(RequestLog log) {
```

Comment:
Ah I see now, prometheus itself has an implicit requirement that labels are recorded in the same order as the label names were initialized. How weird!